File: js/src/browser/pdf.js
/**
 * PDF generation support
 *
 * Wraps Playwright/Puppeteer's page.pdf() method behind a unified interface.
 * Both engines expose identical options: format, printBackground, margin, etc.
 *
 * Note: PDF generation only works in Chromium headless mode.
 * Playwright Firefox and WebKit do not support page.pdf().
 */

import { createEngineAdapter } from '../core/engine-adapter.js';

/**
 * Generate a PDF of the current page
 *
 * @param {Object} options - Configuration options
 * @param {Object} options.page - Playwright or Puppeteer page object
 * @param {string} options.engine - Engine type ('playwright' or 'puppeteer')
 * @param {Object} options.pdfOptions - PDF generation options passed to the engine
 * @param {string} [options.pdfOptions.format] - Paper format (e.g. 'A4', 'Letter')
 * @param {boolean} [options.pdfOptions.printBackground] - Print background graphics
 * @param {Object} [options.pdfOptions.margin] - Page margins { top, right, bottom, left }
 * @param {string} [options.pdfOptions.path] - File path to save the PDF (optional)
 * @param {string} [options.pdfOptions.scale] - Scale of the webpage rendering (0.1–2)
 * @returns {Promise<Buffer>} - PDF as a Buffer
 */
export async function pdf(options = {}) {
  const { page, engine, pdfOptions = {} } = options;
  const adapter = createEngineAdapter(page, engine);
  return await adapter.pdf(pdfOptions);
}

File: python/src/browser_commander/browser/pdf.py
"""PDF generation support.

Wraps Playwright's page.pdf() method behind a unified interface.

Note: PDF generation only works with Playwright in Chromium headless mode.
Selenium/WebDriver does not support native PDF generation.
"""

from __future__ import annotations

from typing import Any

from browser_commander.core.engine_adapter import create_engine_adapter
from browser_commander.core.engine_detection import EngineType


async def pdf(page: Any, engine: EngineType, **options: Any) -> bytes:
    """Generate a PDF of the current page.

    Args:
        page: Playwright page or Selenium WebDriver object
        engine: Engine type ('playwright' or 'selenium')
        **options: PDF generation options forwarded to the underlying engine.
            Common options (Playwright):
            - format: Paper format (e.g. 'A4', 'Letter')
            - print_background: Print background graphics
            - margin: Dict with top/right/bottom/left margins
            - path: Optional file path to save the PDF

    Returns:
        PDF content as bytes

    Raises:
        NotImplementedError: If the engine does not support PDF generation
    """
    adapter = create_engine_adapter(page, engine)
    return await adapter.pdf(**options)
